### PR TITLE
Add navigation component

### PR DESCRIPTION
**Pull request by MancunianSam**

### CODE REVIEW

File: src/nationalarchives/components/nested-navigation/_index.scss [hunk 1]
```diff
@@ -0,0 +1,157 @@
+.govuk-tna-tree {
+  /* Can I use this instead? $govuk-checkboxes-size */
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/_index.scss:2]: Should that comment be here?

File: src/nationalarchives/components/nested-navigation/_index.scss [hunk 2]
```diff
@@ -0,0 +1,157 @@
+.govuk-tna-tree {
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/_index.scss:1]: Is the name convention for the classes correct. At the moment in the frontend sass we've gone with this style: `.govuk-button__tna-button-link`, ie using double underscore. 

I seem to vaguely remember there being a reason for using that style.

File: src/nationalarchives/components/nested-navigation/_index.scss [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  box-shadow: 0 0 0 $govuk-focus-width $govuk-focus-colour;
+}
+
+[role="treeitem"][aria-selected="true"] > .govuk-checkboxes__item > .govuk-checkboxes__label::after,
+[role="treeitem"][aria-selected="true"] > .govuk-tna-tree__node-item__container > .govuk-checkboxes__item > .govuk-checkboxes__label::after {
+  opacity: 1;
+}
+
+/*
+ * Styles pilfered from this thread:
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/_index.scss:143]: Maybe reword this comment to: `Styles copied from this thread...`

File: src/nationalarchives/components/nested-navigation/nested-navigation.test.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+interface NestedNavigationData {
+  name: string
+  type: string
+  id: string
+  children: NestedNavigationData[]
+}
+
+describe('Nested navigation', () => {
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/nested-navigation.test.ts:56]: Should there be a test for when the parent node is selected, al child nodes are selected?

File: src/nationalarchives/components/nested-navigation/nested-navigation.ts [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+  ) => HTMLLIElement[] = (ul, elements) => {
+      for (let i = ul.children.length - 1; i >= 0; i--) {
+        const item: HTMLLIElement | null = ul.children.item(
+          i
+        ) as HTMLLIElement | null
+        if (item !== null) {
+          if (item.nodeName === 'LI') {
+            elements.push(item)
+          }
+          // If children includes a UL/role=group then get children
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/nested-navigation.ts:148]: Maybe re-word comment slightly: `If children include a ...`

File: src/nationalarchives/components/nested-navigation/nested-navigation.ts [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+            if (parent?.nextElementSibling != null) {
+              this.setFocusToItem(parent.nextElementSibling as HTMLLIElement)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  handleKeyDown: (ev: KeyboardEvent) => void = (ev) => {
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/nested-navigation.ts:296]: This is a pretty long function, with some complex logic for some of the cases. Is it possible to move the case logic to separate functions and call those? Not sure that would make things better or worse?

File: src/nationalarchives/components/nested-navigation/nested-navigation.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+          type: node
+          id: folderid12
+          children:
+            - name: file12
+              type: item
+              id: fileid12
+            - name: file22
+              type: item
+              id: fileid22
+        - name: file1
```
Comment (TomJKing) [inline src/nationalarchives/components/nested-navigation/nested-navigation.yaml:32]: Is having the same file name twice in different folders deliberate?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just a few very minor things.

File: jest-puppeteer.config.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// jest-puppeteer.config.js
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply jest-puppeteer.config.js:1]: No, I copied the file from the govuk repo. I've removed it.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,65 @@
+{
+  "name": "@nationalarchives/tdr-components",
+  "version": "1.0.11",
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply package.json:3]: Yep, this is just showing how many failed deploys I had. I've set it back to 1.0.0

File: src/nationalarchives/components/nested-navigation/_index.scss (hunk 1 above)
Comment (MancunianSam) [reply src/nationalarchives/components/nested-navigation/_index.scss:2]: I don't think so, I've removed it.

File: src/nationalarchives/components/nested-navigation/_index.scss (hunk 2 above)
Comment (MancunianSam) [reply src/nationalarchives/components/nested-navigation/_index.scss:1]: We are doing that sort of style though, there's things like `govuk-tna-tree__node-item__container` I'm not sure though, best to check with @garethfoote

File: src/nationalarchives/components/nested-navigation/nested-navigation.test.ts (hunk 4 above)
Comment (MancunianSam) [reply src/nationalarchives/components/nested-navigation/nested-navigation.test.ts:56]: Yeah, that makes sense, I've added one in.

File: src/nationalarchives/components/nested-navigation/nested-navigation.ts (hunk 6 above)
Comment (MancunianSam) [reply src/nationalarchives/components/nested-navigation/nested-navigation.ts:296]: I think your right on the ArrowRight, ArrowLeft and End ones so I've moved those into separate private functions. The others are two lines so I think that's ok.

File: src/nationalarchives/components/nested-navigation/nested-navigation.yaml (hunk 7 above)
Comment (MancunianSam) [reply src/nationalarchives/components/nested-navigation/nested-navigation.yaml:32]: No, just lazy typing, I've changed it to something different and updated the tests.

File: package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (TomJKing) [reply package.json]: Not sure this version roll back works, given the other PR you had to do?

[the remark this answers is not in the document]
Comment (MancunianSam) [reply package.json]: This is the "dev" version of the library and isn't needed at all. If you look at the way GDS have done it, [their main package.json](https://github.com/alphagov/govuk-frontend/blob/main/package.json) doesn't have a version as this is never published, the [file in the package directory](https://github.com/alphagov/govuk-frontend/blob/main/package/package.json) has the version which gets published. 

It's a similar thing here, I've bumped the version in `package/package.json` but it is confusing having two versions so I've done what GDS have done and removed `version` from the package.json in the root directory.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
